### PR TITLE
Fixes for timing and clang-tidy use.

### LINKED ISSUE #28: Timing calculations are wrong
There are a few issues with the timing calculations:

* A rounding / type conversion error can lead to *wildly* incorrect values (I had one report that came up at 7 seconds, but really took something on the order of a few dozen milliseconds).  This was seen on Windows.  Preserving the values as nanoseconds and only doing a division at the end gives better results.  (And it appears that dividing by 1e9 is better than multiplying by 1e-9 if you want to avoid loss due to casting bugs.)

*  The Linux MONONTONIC_RAW clock should not really be used.  Besides not being portable, the values are not guaranteed to be microseconds -- and really you do want any adjtime adjustments because those reflect real-world timekeeping.  (This makes it not as good for reproducible benchmarking, but the cpu clock is better for that anyway.)

* There is an order of magnitude bug in the POSIX timings, leading to times being reported as 1/10th their actual value (this lead to a lot of confusion for me at one point).  Hint, 1 billionth is not 10e-9, but rather 1e-9.

I have a PR coming that addresses these.

### LINKED ISSUE #29: TEST_CASE should evaluate as a statement (no trailing semicolon)
Clang-Tidy found this.

TEST_CASE(name) ends with a semicolon, which means that if you use it like this:

TEST_CASE(x == 1);

The compiler complains about an empty statement (caused by the redundant semicolon).  This can also lead to surprises if this is used in contexts without enclosing braces, so it's not recommended.

Simple solution is just to elide the trailing semicolon.

PR with this is on the way.

### PR DESCRIPTION
**Pull request by gdamore**
fixes #28 Timing calculations are wrong
fixes #29 TEST_CASE should evaluate as a statement (no trailing semicolon)

I've added my copyright as a "Portions" bit.  If you prefer to instead just have me add a Copyright line below yours (how I do it for my repos), please let me know (or just make the change yourself).

Meaning, I'm also happy with:

```
 * Copyright (c) 2013-2019 Martin Mitas
 * Copyright 2019 Garrett D'Amore 
 *
 * ... license text follows
```

Oh, and thanks for your work on this project -- I'm now using it in NNG (as you know), and have begin replacing my own framework (c-convey -- see github.com/gdamore/c-convey) with it as this is simply easier to use and debug.

As an aside, the "(c)" in your copyright statement is actually useless (following a bad practice started by companies years ago who should have known better).   Because (c) is not recognized as the copyright symbol, and even if it were, it is entirely redundant to the word "Copyright" in copyright law.  (The symbol was intended for uses where a shorthand substitute for the word "Copyright" was desired.)

### CODE REVIEW

**Comment by mity** [comment]
> I've added my copyright as a "Portions" bit. If you prefer to instead just have me add a Copyright line below yours (how I do it for my repos), please let me know (or just make the change yourself).

Yes please. That would be much better.

**Comment by gdamore** [comment]
> > I've added my copyright as a "Portions" bit. If you prefer to instead just have me add a Copyright line below yours (how I do it for my repos), please let me know (or just make the change yourself).
> 
> Yes please. That would be much better.

Done.

File: include/acutest.h [hunk 1]
```diff
@@ -384,12 +385,7 @@ static jmp_buf test_abort_jmp_buf__;
     test_timer_init__(void)
     {
         if(test_timer__ == 1)
-    #ifdef CLOCK_MONOTONIC_RAW
```
Comment (mity) [inline include/acutest.h]: Could you please explain this more? 

My motivation was to provide an elapsed time as with a stopwatch, without some calendar events like e.g. leap seconds, switches to summer time zones, and especially NTP syncing to interfere, if the system is capable of that.

But I admit I am no expert at Linux timer so maybe I did the direct opposite?

File: include/acutest.h [hunk 2]
```diff
@@ -178,7 +179,7 @@
  * the last test case after exiting the loop), you may use TEST_CASE(NULL).
  */
 #define TEST_CASE_(...)         test_case__(__VA_ARGS__)
-#define TEST_CASE(name)         test_case__("%s", name);
+#define TEST_CASE(name)         test_case__("%s", name)
```
Comment (mity) [inline include/acutest.h]: Yeah. Stupid typo. Wondering how it could survive for so long.

File: include/acutest.h (hunk 1 above)
Comment (gdamore) [reply include/acutest.h]: Historically, CLOCK_MONOTONIC_RAW was provided to give a faster way to get fine grained performance numbers -- there is math done for adjtime() -- and also if you wanted to have timing data that was subject to modifications by NTP.

The thing is, NTP is actually supposed to be making those adjustments for a reason -- and that is that processor clocks drift (they're pretty subject to changes due to thermal conditions for example).

Generally this mostly means that if you want to count cycles, CLOCK_MONOTONIC_RAW might be what you want.  But most of the time, you want real-world time, which is what CLOCK_MONOTONIC is for, although there is also CLOCK_REALTIME, but the problem with CLOCK_REALTIME is that it is subject to major clock changes e.g. due to timezone, etc. and can even go backwards.

CLOCK_MONOTONIC should mostly give the same amount of real-world time for each quanta *because* it is adjusting for differences in the system clock by synchronizing with external timing sources (assuming you have a working NTP implementation).

Basically, my recommendation is never to use CLOCK_MONOTONIC_RAW unless you know you have a specific requirement to avoid the minor slewing that NTP makes via adjtime().  The only times that requirement really comes up is if you're implementing another clock synchronization system (a different ntpd for example) or you want to count *cycles* irrespective of vagaries in the system clock frequency.

Mostly this means, avoid CLOCK_MONOTONIC_RAW for ~everyone.

Comment (gdamore) [reply include/acutest.h]: Btw, CLOCK_MONOTONIC will not be affected by timezones, leap seconds, or major events like ntpdate running to reset the clock.  It will only be affected by *very* small frequency adjustments made by NTP to accommodate clock drift.  (I forget the limit it will make, but the frequency change is a small percentage.)

Comment (mity) [reply include/acutest.h]: > The thing is, NTP is actually supposed to be making those adjustments for a reason -- and that is that processor clocks drift (they're pretty subject to changes due to thermal conditions for example).

I understand that.

My thinking was going as this: I assume the drift may accumulate for a long time period to reach some substantial gap from `CLOCK_MONOTONIC`. But then, when it does, the adjustment can occur exactly in the very short time period while the test is being executed. The result would then be likely more off then without the adjustment. And I guess the vast majority of tests is likely executed in < 1 sec time frame.

Is my assumption wrong and the NTP is synced so often this plays no real role?

Comment (mity) [reply include/acutest.h]: Oops. Race condition. Your latest comment seen after page refresh explains that.

Comment (gdamore) [reply include/acutest.h]: Yes, you're misunderstanding how this works.  The concerns you have would apply to other clocks, but not CLOCK_MONOTONIC.

NTP via adjtime adjusts only the clock *frequency*, and only by small amounts.  This has a few ramifications:

1. The monotonic nature (always increasing a single microsecond at a time) is preserved.
2. You will never see a large difference in the real-world time between one sampling of CLOCK_MONOTONIC and the next.

**Comment by mity** [comment]
Merged. Thanks a lot for the fixes and the explanation.
